[4-6단계 - 방탈출 사용자 예약] 라젤(문선영) 미션 제출합니다.

RaZel713 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [x] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

RaZel713 committed:
refactor: 누락된 접근 제한자 구현

RaZel713 committed:
refactor: CreateReservationRequest import 추가

RaZel713 committed:
docs: README 정리

RaZel713 committed:
feat: 로그인 페이지 컨트롤러 구현

RaZel713 committed:
feat: 멤버 테이블 스키마 정의

RaZel713 committed:
feat: 로그인 기능 구현

RaZel713 committed:
feat: 인증 정보 조회 기능 구현

RaZel713 committed:
feat: 로그인 기능 구현에 따른 html 추가

RaZel713 committed:
test: 로그인 기능 http 파일 구현

RaZel713 committed:
feat: 멤버 데이터 추가

RaZel713 committed:
refactor: 패키지 구조 변경

RaZel713 committed:
feat: 예약 테이블 스키마 정의

- 멤버 테이블을 외래키로 추가

RaZel713 committed:
refactor: DTO 네이밍 변경

RaZel713 committed:
feat: 멤버 도메인 유효성 검사 구현

RaZel713 committed:
feat: 예약 생성 시, 로그인한 사용자 정보를 활용

RaZel713 committed:
refactor: 예약 데이터 수정

RaZel713 committed:
feat: 관리자 예약 생성 시, 유저를 조회해 선택 후 생성

RaZel713 committed:
refactor: 패키지 구조 변경

RaZel713 committed:
refactor: 멤버 도메인 네이밍 변경

RaZel713 committed:
feat: 접근 권한 제어 기능 구현

RaZel713 committed:
feat: 예약 목록 검색 기능 구현

RaZel713 committed:
test: 테스트 패키지 구조 변경

- 프로덕션 코드의 패키지 구조와 일치하도록 테스트 패키지를 정리한다

RaZel713 committed:
refactor: 도메인 명과 데이터 테이블 명이 일치하도록 변경

RaZel713 committed:
test: 로그인 구현에 따른 예약 테스트 수정

RaZel713 committed:
test: 접근 권한 제어 기능 구현에 따른 도메인 페이지 테스트 수정

RaZel713 committed:
refactor: 메서드 네이밍 변경

RaZel713 committed:
refactor: 응답 DTO에서 비밀 번호 삭제

RaZel713 committed:
test: 로그인 페이지 컨트롤러 테스트 추가

RaZel713 committed:
test: 로그인 컨트롤러 테스트 추가

RaZel713 committed:
refactor: 응답 DTO 수정

RaZel713 committed:
fix: 어드민 페이지에서 예약 추가 오류 수정

jinyoungchoi95 commented on build.gradle:
```diff
@@ -26,6 +26,10 @@ dependencies {
     testImplementation 'org.springframework.boot:spring-boot-starter-test'
     testImplementation 'io.rest-assured:rest-assured:5.3.1'
+
+    implementation 'io.jsonwebtoken:jjwt-api:0.11.5'
```
jjwt를 사용한 이유가 있을까요? 메이저 버전이 0인데 믿고 사용할 수 있는 라이브러리일까요?

RaZel713 replied:
jjwt는 2015년부터 시작된 오픈 소스 라이브러리로, 널리 사용되고 있습니다.
현재 메이저 버전은 존재하지 않지만, 활발한 유지 관리와 커뮤니티 기여를 통해 안정적으로 사용되고 있는 라이브러리라고 생각합니다.

다만, 신뢰성이나 장기 유지보수를 더 중요하게 여긴다면 
`auth0/java-jwt` 와 같이 많이 사용되고, 메이저 버전이 존재하는 라이브러리를 사용할 수 있을 것 같습니다.

출처: 
[Maven Repository jjwt api](https://mvnrepository.com/artifact/io.jsonwebtoken/jjwt-api)
[jjwt github](https://github.com/jwtk/jjwt)

jinyoungchoi95 replied:
활발하다라는 것만으로는 도입하는데 무리가 있을 것 같아요. major 버전이 0이라는 것은 대다수의 개발자가 신뢰하지 못하게 만드는 지표입니다. major 버전이란 것은 그만큼 신뢰도있게 소스를 제공하고 있냐를 메인테이너가 소개하는 지표이고, 0이란 것은 아직도 본인들이 신뢰도가 떨어지는 것이 있다라는 것을 의미하는 것이니까요.

제가 도입해야하는 라이브러리를 찾아야하는 상황이었다면 최우선 필터링조건으로 major 버전을 봤을 것 같네요.

https://github.com/jwtk/jjwt/discussions/900

RaZel713 replied:
`jjwt` 대신 `auth0/java-jwt` 라이브러리를 사용하도록 코드를 변경하였습니다!

Q) 오리의 최우선 필터링 조건이 major 버전이라고 하셨는데, 혹시 다른 필터링 조건도 여쭈어 봐도 될까요?
제가 찾아보기로는

- 최근 커밋/릴리즈 시점 (유지 보수가 이루어지고 있는지?)
- 공식 문서 / 사용 예제 
- 테스트 커버리지 (신뢰 가능한가?)
- 질문과 답변이 활발한가? (깃허브 등 커뮤니티)
- CVE 등재 여부 (보안 이슈)
- 라이브러리 크기 / 종속성 수 (너무 많은 종속성을 갖고 있는가?)

이러한 필터링 조건들을 고려한다고 하는데, 
현 시점에서 어떠한 점들을 최우선적으로 봐야한 지 궁금합니다.

jinyoungchoi95 replied:
- 내가 필요한 기능이 정확하게 있는가 
- 앞으로도 추가 기능이 필요하다면 그런 기능들도 쓸 수 있게 지원되어있는가
- 메이저 버전의 업데이트가 활발한가 (굵직한 기능들을 계속 추가중인가)
- 사내에서 사용중인 오픈소스인가
- 내가 혹은 팀에서 경험해본 오픈소스인가

등의 이유가 있겠네요.

jinyoungchoi95 commented on src/main/java/roomescape/auth/controller/LoginController.java:
```diff
@@ -0,0 +1,45 @@
+        CreateTokenServiceRequest request = CreateTokenServiceRequest.from(loginRequest);
+
+        String token = authService.createToken(request);
+        ResponseCookie cookie = authService.createCookie(token);
```
`Service`가 web계층에서 사용되는 Cookie를 처리해도되나요?

RaZel713 replied:
`CookieHandler` 유틸 클래스로 쿠키 생성 및 추출 로직을 분리하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/auth/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,86 @@
+
+@Component
+public class JwtTokenProvider {
+    @Value("${security.jwt.token.secret-key}")
+    private String secretKey;
+    @Value("${security.jwt.token.expire-length}")
+    private long validityInMilliseconds;
+
+    private Key key;
+
+    @PostConstruct
+    public void init() {
+        byte[] keyBytes = Base64.getDecoder().decode(secretKey);
+
+        this.key = Keys.hmacShaKeyFor(keyBytes);
```
`@Value`는 생성자의 파라메터에서 사용될 수 있으니 `@PostConstruct`를 제거하고 불변하게 만들어보세요.

https://docs.spring.io/spring-framework/reference/core/beans/annotation-config/value-annotations.html

RaZel713 replied:
```java
private final Key key;
private final long validityInMilliseconds;

public JwtTokenProvider(
            @Value("${security.jwt.token.secret-key}") String secretKey,
            @Value("${security.jwt.token.expire-length}") long validityInMilliseconds
    ) {
        byte[] keyBytes = Base64.getDecoder().decode(secretKey);
        this.key = Keys.hmacShaKeyFor(keyBytes);
        this.validityInMilliseconds = validityInMilliseconds;
}
```
생성자 파라미터에 `@Value`를 사용하여 `secretKey`와 `validityInMilliseconds` 값을 주입받고, 
이를 불변 필드로 초기화하게끔 수정하였습니다!

jinyoungchoi95 commented on src/main/java/roomescape/auth/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,86 @@
+                .setClaims(claims)
+                .claim("name", member.getName())
+                .claim("role", member.getRole())
+                .setIssuedAt(now)
```
issuedAt이 현재 사용되는 곳이 없는데 필요한 정보인가요?

RaZel713 replied:
issuedAt은 JWT 토큰의 발급 시점을 나타내지만, 
현재는 `.setExpiration(validity)`를 통해 만료 시간(exp)만으로 유효성을 검증하고 있어 
불필요하다고 판단하여 해당 클레임을 제거하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/auth/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,86 @@
+                .claim("role", member.getRole())
+                .setIssuedAt(now)
+                .setExpiration(validity)
+                .signWith(key, SignatureAlgorithm.HS256)
```
`HS256`를 사용한 이유가 있나요?

RaZel713 replied:
학습 테스트를 하는 과정에 있어, `SignatureAlgorithm.HS256`이 
서명을 위한 알고리즘이라는 것, 가장 많이 사용되는 안정적인 알고리즘이라는 것 정도만 알고 사용하였습니다!

오리의 피드백을 바탕으로 추가 학습의 필요성을 느껴 학습 후 내용을 간략하게 올립니다.

## HS256 이란?
HS256(HMAC with SHA-256)은 JWT(Json Web Token)에서 널리 사용되는 **서명 알고리즘** 입니다.

- `HS256`은 대칭키 알고리즘으로, 하나의 비밀 키만을 사용해 서명하고 검증합니다. 
- `SHA-256` 기반의 해싱 알고리즘은 비교적 속도가 빠르고 연산 부담이 적습니다. 

#### JWT에서 HS256의 역할
- JWT는 Base64로 인코딩된 Header, Payload, Signature(해시 값)로 구성됩니다.
- HS256은 Header + Payload에 대해 비밀키로 HMAC-SHA256 서명을 생성하여 Signature에 넣습니다.
- 이 서명은 “토큰이 위변조되지 않았다”는 것을 검증할 수 있도록 도와줍니다.

<details>
<summary> 📕 추가 학습) 암호화와 서명 </summary> 

**암호화 (Encryption)** 의 목적은 내용을 숨기는 것 (기밀성)입니다.
**서명 (Signature)** 의 목적은 내용이 위조되지 않았음을 증명하는 것(무결성 + 인증)입니다.

따라서 HS256은 암호화 알고리즘이 아닌 서명 알고리즘이라고 할 수 있습니다.

</details>

질문하신 의도에 맞게 답변한 걸까요..?

jinyoungchoi95 replied:
네 맞습니다. HS256 말고 다른것들은 없는지 한번 학습해보시면 도움될 것 같아요.

RaZel713 replied:
## ⁉️ 서명을 위한 알고리즘은 어떤 것이 있는가?

### jwt가 지원하는 서명 알고리즘

**1.  HMAC (Hash-based Message Authentication Code)**
- 공유 비밀 키 기반 서명 방식 (대칭 키)
- 종류: 사용하는 해시 함수 (SHA-256, SHA-384, SHA-512) 에 따라 HS256, HS384, HS512가 존재합니다.
✅ 장점: 빠르고 구현이 간단합니다
⚠️ 단점: 비밀 키 유출 시 모든 노드가 위험합니다 (하나의 키를 모두 공유하기 때문)

**2. RSA (Rivest–Shamir–Adleman)**
- 비대칭 키 기반 서명 방식 (공개 키 & 개인 키)
- 종류: 사용하는 해시 함수 (SHA-256, SHA-384, SHA-512) 에 따라 RS256, RS384, RS512가 존재합니다.
✅ 장점: 개인 키로 서명하고 공개 키로 검증 가능합니다 → 즉, 인증 서버와 리소스 서버가 분리 가능해 집니다.
⚠️ 단점: HMAC보다 속도 느리고, 키 길이/관리 복잡합니다.

**3. ECDSA (Elliptic Curve Digital Signature Algorithm)**
- 비대칭 키 기반 서명 방식 + 더 짧은 키로 같은 수준의 보안 제공 합니다
- 종류: 사용하는 해시 함수 (SHA-256, SHA-384, SHA-512) 에 따라 ES256, ES384, ES512가 존재합니다.
✅ 장점: RSA보다 작은 키 크기로 같은 보안 강도 제공합니다.
⚠️ 단점: 구현이 복잡하고 일부 플랫폼에서 호환성 이슈가 존재합니다.

---

간단한 내부 인증이 필요하고, 빠른 성능 필요한 경우 : HS256
서비스 간 공개 키 기반 검증 필요한 경우 : RS256
더 강력한 보안과 작은 메시지 크기가 필요한 경우 (모바일 환경이나 네트워크 최적화가 필요한 경우) : ES256

<details>
<summary> 📕 DEEP DIVE) RS256 보다 ES256가 더 강력한 보안인 이유 </summary>

1. **짧은 키로도 높은 보안성 확보**  
   - ES256은 타원 곡선 암호학(Elliptic Curve Cryptography, ECC) 기반입니다.
   - ECC 알고리즘의 수학적 구조 덕분에, 짧은 키로도 RSA보다 강력한 보안 제공이 가능합니다.
   - RSA의 보안 강도는 112-bit이고, ECC (ES256)의 보안 강도는 128-bit 입니다.
   - 더 작은 키는 연산량도 줄여줄 수 있습니다.

2. **서명 크기가 작아 전체 JWT 크기 축소**  
   - ES256의 서명은 약 64바이트로, RS256보다 훨씬 작습니다.
   - 네트워크 전송 및 저장 공간 측면에서 효율적입니다.

3. **성능이 뛰어남**  
   - 계산 복잡도는 낮고 처리 속도는 빠릅니다.
   - 특히 모바일, IoT 같은 리소스 제한 환경에 적합합니다.

</details>

참고: JWT Handbook

jinyoungchoi95 commented on src/main/java/roomescape/auth/service/AuthService.java:
```diff
@@ -0,0 +1,76 @@
+public class AuthService {
+
+    private final MemberDao memberDao;
+    private final JwtTokenProvider jwtTokenProvider;
```
`JwtTokenProvider`는 jjwt라는 외부 의존성이 걸린 구현체인데 구현체에 의존해도 괜찮나요?

RaZel713 replied:
### JwtTokenProvider 에 의존하는 경우 발생하는 문제
- `JwtTokenProvider`는 jjwt라는 외부 라이브러리에 의존하고 있기 때문에, 
해당 구현체를 서비스 레이어에서 직접 사용하는 것은 외부 기술에 강하게 결합되는 구조가 됩니다.
- 유지 보수나 구현체 교체에 있어 불리하게 작용합니다.
- 테스트하기에 어렵습니다.

위의 문제점을 해결하기 위해
`AuthTokenProvider`라는 인터페이스를 정의하고, `JwtTokenProvider`가 이를 구현하게 한 뒤, 
서비스에서는 인터페이스에 의존하도록 리팩터링하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/auth/service/AuthService.java:
```diff
@@ -0,0 +1,76 @@
+    private final MemberDao memberDao;
+    private final JwtTokenProvider jwtTokenProvider;
+
+    public AuthService(final MemberDao memberDao, JwtTokenProvider jwtTokenProvider) {
```
final이 달린곳과 달리지 않은 곳이 혼재되어있는데, final을 모두 달것이 아니라면 일관성을 위해 final은 제거하는 것이 좋습니다.

RaZel713 replied:
final 키워드가 빠진 부분들을 수정하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/auth/service/AuthService.java:
```diff
@@ -0,0 +1,76 @@
+                .build();
+    }
+
+    public String extractTokenFromCookie(final Cookie[] cookies) {
```
위와 마찬가지로 intercepter에서 사용되더라도 cookie는 web계층에서 사용되는데 service가 web계층에 의존해도 되나요?

RaZel713 replied:
위의 [피드백](https://github.com/woowacourse/spring-roomescape-member/pull/267#discussion_r2083218570) 에 답변하였습니다!

jinyoungchoi95 commented on src/main/java/roomescape/auth/CheckLoginInterceptor.java:
```diff
@@ -0,0 +1,35 @@
+
+        final Cookie[] cookies = request.getCookies();
+
+        final String token = authService.extractTokenFromCookie(cookies);
+        Member member = authService.findMember(token);
+
+        if (member == null || !member.getRole().equals("Admin")) {
```
1. token에 role이 담겨있는데 member의 role을 검증할 필요가 있나요?
2. token으로 이미 유저에 대해 검증할 수 있는데 member == null이 필요한가요? 즉, 토큰 검증 외의 `authService.findMember` 동작이 필요한가요?

RaZel713 replied:
token을 사용하여 유저에 대해 검증하고, token에서 role을 추출하여
어드민 여부를 검증하는 방식으로 수정하였습니다.

```java
if (!authService.isAdmin(token)) {
            response.setStatus(401);
            return false;
}

public boolean isAdmin(final String token) {
        return authTokenExtractor.extractMemberRoleFromToken(token).equals("ADMIN");
}

@Override
    public String extractMemberRoleFromToken(String token) {
        validateValidToken(token);

        return Jwts.parserBuilder()
                .setSigningKey(key)
                .build()
                .parseClaimsJws(token)
                .getBody()
                .get("role", String.class);
}
```

jinyoungchoi95 commented on src/main/java/roomescape/auth/CheckLoginInterceptor.java:
```diff
@@ -0,0 +1,35 @@
+        Member member = authService.findMember(token);
+
+        if (member == null || !member.getRole().equals("Admin")) {
+            response.setStatus(401);
```
인증과 인가의 개념에 대해서 학습해보세요.

RaZel713 replied:
##  인증(Authentication)과 인가(Authorization)의 개념

###  인증 (Authentication) 
- 시스템이 정당한 사용자인지 판단하기 위해 사용자가 누구인지 신원을 확인하는 과정이다.
- 방법:
  - ID/비밀번호
  - OAuth 로그인(Google, Kakao 등)
  - 생체 정보(지문, 얼굴)
  - 토큰 기반 인증 (JWT 등)

⚪️ 예시:
사용자가 이메일과 비밀번호를 입력하여 로그인 → 서버가 일치하는 사용자 정보를 찾음 → 로그인 성공 시 토큰 발급

---

###  인가 (Authorization) 
- 사용자가 허용된 리소스에만 접근하도록 제한하고, 인증된 사용자가 어떤 권한을 가졌는지 결정하는 과정이다.
- 방법:
  - 권한(role): ADMIN, USER
  - 권한에 따른 API 접근 제어
  - URI, 메서드 기반 접근 제한

⚪️ 예시:
로그인한 사용자가 관리자 페이지 /admin에 접근 → 서버는 해당 사용자의 권한이 ADMIN인지 확인 → 아니라면 접근 거부 (403 Forbidden)

---

### 🔁 둘의 관계
- 인증이 먼저, 인가는 그 다음에 일어난다.
- 인증 없이 인가는 불가능하다.

jinyoungchoi95 commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -0,0 +1,93 @@
+    private static final int MAX_LENGTH = 255;
+
+    private final long id;
+    private final String role;
```
role을 string이 아닌 enum으로 관리할 수 있지 않을까요?

RaZel713 replied:
enum을 사용하여 role을 관리할 수 있도록 수정하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -0,0 +1,93 @@
+    private final long id;
+    private final String role;
+    private final String name;
+    private final String email;
```
이메일 형식에 대한 검증은 필요없을까요?

RaZel713 replied:
이메일 형식에 대한 검증을 추가하였습니다.

```java
private static final String EMAIL_REGEX = "^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$";
private static final Pattern EMAIL_PATTERN = Pattern.compile(EMAIL_REGEX);

 private void validateEmail(final String email) {
        if (!EMAIL_PATTERN.matcher(email).matches()) {
            throw new InvalidInputException("올바르지 않은 이메일 형식입니다: " + email);
        }
}
```

jinyoungchoi95 commented on src/main/resources/application.properties:
```diff
@@ -3,3 +3,5 @@ spring.h2.console.path=/h2-console
 spring.datasource.username=test
 spring.datasource.password=
+security.jwt.token.secret-key=Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=
+security.jwt.token.expire-length=3600000
```
시간의 단위를 명시하여야 어떻게 사용되는 지 알 수 있지 않을까요?

RaZel713 replied:
```
security.jwt.token.expire-length=3600000 # 1시간 (1000 * 60 * 60 밀리초)
```

주석으로 시간 단위를 명시하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/auth/service/AuthService.java:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+    public Member findMember(final String token) {
```
무엇을 가지고 member를 찾는 건지 메서드 네이밍에 명시되어야하지 않을까요?

RaZel713 replied:
`findMemberByToken` 이라고 명명하여
토큰을 기준으로 member를 찾는 메서드 임을 나타내었습니다.

jinyoungchoi95 commented on src/main/java/roomescape/auth/service/AuthService.java:
```diff
@@ -0,0 +1,76 @@
+import roomescape.member.domain.Member;
+
+@Service
+public class AuthService {
```
dm으로 먼저 말씀주셨듯 테스트가 전반적으로 부족하니 추가해주세요.

RaZel713 replied:
인증 기능 관련 테스트를 추가하였습니다!

jinyoungchoi95 reviewed (changes requested):
안녕하세요 라젤 😄

몇가지 코멘트 남겨두었으니 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

RaZel713 committed:
feat: 멤버 도메인에 이메일 형식 검증 로직 구현

RaZel713 committed:
refactor: final 키워드 추가

RaZel713 committed:
feat: 인증 만료 기한 단위를 주석으로 명시

RaZel713 committed:
refactor: 메서드 명 수정

- 인증 서비스의 'findMember' 메서드 명을 'findMemberByToken'로 변경하여 토큰을 기준으로 멤버를 찾는 메서드 임을 명시한다.

RaZel713 committed:
refactor: admin 검증을 interceptor에서 service로 위임

RaZel713 committed:
refactor: AuthService가 구현체가 아닌 인터페이스에 의존하도록 리팩터링

- jjwt 외부 라이브러리에 대한 강한 결합을 줄이기 위해 AuthTokenProvider 인터페이스를 도입한다.
- JwtTokenProvider는 AuthTokenProvider 구현체로 변경한다.

RaZel713 committed:
refactor: 토큰 생성과 추출 책임을 AuthTokenProvider와 AuthTokenExtractor로 분리

RaZel713 committed:
refactor: AuthTokenProvider와 AuthTokenExtractor의 메서드 이름을 역할에 맞게 변경

RaZel713 committed:
refactor: @Value를 생성자 파라미터에 적용하여 의존성 주입

- secretKey와 validityInMilliseconds 값을 생성자 파라미터에서 주입받아 불변 필드로 초기화한다.

RaZel713 committed:
refactor: 불필요한 issuedAt 클레임 제거

RaZel713 committed:
refactor: 멤버 역할을 문자열 대신 Role enum으로 표현

RaZel713 committed:
refactor: 유틸 클래스로 쿠키 생성 및 추출 로직 분리

RaZel713 committed:
refactor: 컨트롤러 필드명 변경

RaZel713 committed:
refactor: 멤버 도메인의 이름, 이메일, 비밀 번호 길이 제한 변경

RaZel713 committed:
test: 멤버 도메인 테스트 구현

RaZel713 committed:
feat: 회원 가입 기능 구현

RaZel713 committed:
refactor: 누락된 접근 제한자 추가

RaZel713 committed:
test: 테스트 데이터 형식 수정

RaZel713 committed:
feat: 로그 아웃 기능 구현

RaZel713 committed:
docs: README 수정

RaZel713 committed:
test: 회원 가입 페이지 테스트 추가

RaZel713 committed:
test: 멤버 컨트롤러 테스트 추가

RaZel713 committed:
test: 멤버 서비스 테스트 추가

RaZel713 committed:
test: 멤버 DAO 테스트 추가

RaZel713 committed:
refactor: interceptor 패키지 생성

RaZel713 committed:
test: JwtTokenExtractor 테스트 추가

RaZel713 committed:
test: JwtTokenProvider 테스트 추가

RaZel713 committed:
test: CookieHandler 테스트 추가

RaZel713 committed:
test: 로그아웃 테스트 추가

RaZel713 committed:
test: 인증 서비스 테스트 추가

RaZel713 reviewed (commented):
안녕하세요 ~오리! 
리뷰를 열심히 반영해 보았지만, 
저번에 리뷰 주셨던 bdd 주석 관련 내용은 제대로 반영하지 못했습니다 🥲
해당 부분은 이번 리뷰 이후에 반영하여 올리도록 하겠습니다 

추가로, 저번 1-3단계 마지막 피드백 중 답변 드리지 못한 부분이 있어 답변 드립니다.

>ResponseEntity에서 제공하는 기본 빌더 메서드를 사용하지 않는 이유가 있을까요?
> [ 1-3단계 리뷰 링크](https://github.com/woowacourse/spring-roomescape-member/pull/169#discussion_r2077443949)

➡️  기본 빌더 메서드에 conflict 가 존재하지 않아, 코드의 일관성을 위해 모두 아래와 같은 형태로 작성하였습니다.
``` java
@ExceptionHandler
    public ResponseEntity<String> handleInUseException(InUseException e) {
        return ResponseEntity.status(HttpStatus.CONFLICT).body(ERROR_PREFIX + e.getMessage());
}
```

RaZel713 committed:
refactor: 로그인 체크 응답 DTO 네이밍 변경

RaZel713 committed:
refactor: 컨트롤러 요청 DTO에서 서비스 요청 DTO를 생성할 수 있도록 변경

jinyoungchoi95 commented on src/main/java/roomescape/auth/resolver/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,39 @@
+import roomescape.auth.service.AuthService;
+import roomescape.auth.util.CookieHandler;
+
+public class LoginMemberArgumentResolver implements HandlerMethodArgumentResolver {
```
`Resolver`나 `Interceptor`는 테스트가 필요없을까요?

RaZel713 replied:
`Resolver`와 `Interceptor`는 컨트롤러 실행 전에 동작하며, 주로 요청 전처리나 인증/인가와 같은 역할을 담당합니다.

컨트롤러 실행 전 공통적으로 동작하는 로직이기 때문에, 
다른 컨트롤러 테스트들을 통해 간접적으로 검증할 수 있다고 판단해 별도의 직접적인 테스트는 필요하지 않다고 생각합니다.

<img width="672" alt="스크린샷 2025-05-16 오전 9 34 38" src="https://github.com/user-attachments/assets/1fabcadc-3e0e-49c9-bdfb-0955ade4e914" />

jinyoungchoi95 replied:
다른 테스트에 의해서 해당 클래스가 커버되므로 테스트를 작성하지 않는 것에 대해서는 다시 생각해보세요.

다른 테스트가 interceptor를 커버하지 않는다면 interceptor를 새로 사용하는 클래스 입장에서는 무엇을 믿고 사용할 수 있을까요? 현재와 같은 기조라면 controller에서 간접적으로 테스트되고 있으니 service, repository, domain 모두 테스트하지 않아도 된다로 이어질 수도 있겠네요.

jinyoungchoi95 commented on src/main/java/roomescape/auth/interceptor/CheckLoginInterceptor.java:
```diff
@@ -0,0 +1,33 @@
+        final Cookie[] cookies = request.getCookies();
+        final String token = CookieHandler.extractTokenFromCookies(cookies);
+
+        if (!authService.isAdmin(token)) {
+            response.setStatus(401);
```
> 사용자가 허용된 리소스에만 접근하도록 제한하고, 인증된 사용자가 어떤 권한을 가졌는지 결정하는 과정이다.

답변주신대로라면 어드민인지 검증하여 반환하는 것이 401(인증)이 맞나요?

말씀하신 기준이라면 CheckLogin은 인증이고, isAdmin은 인가이고, 401은 인증인데 혼용되어있네요.

RaZel713 replied:
- AuthenticationException(`401 UNAUTHORIZED`) : 사용자 인증 예외
- AuthorizationException(`403 FORBIDDEN`): 사용자 인가 예외

로 예외를 나누어 표기하도록 변경하였습니다.

jinyoungchoi95 commented on src/main/java/roomescape/auth/service/AuthService.java:
```diff
@@ -0,0 +1,56 @@
+        return authTokenProvider.createTokenFromMember(member);
+    }
+
+    private Member findMemberByEmail(final String email) {
+        return memberDao.findByEmail(email)
+                .orElseThrow(() -> new AuthorizationException("존재하지 않는 이메일 입니다"));
+    }
```
재사용되지 않고, 라인이 한줄인 메서드는 불필요한 메서드 래핑인 것 같아요.

RaZel713 replied:
해당 메서드를 제거하였습니다!

jinyoungchoi95 commented on src/main/java/roomescape/auth/service/AuthService.java:
```diff
@@ -0,0 +1,56 @@
+                .orElseThrow(() -> new AuthorizationException("존재하지 않는 이메일 입니다"));
+    }
+
+    public Member findMemberByToken(final String token) {
+        validateTokenExisted(token);
+        final long memberId = Long.parseLong(authTokenExtractor.extractMemberIdFromToken(token));
+
+        return memberDao.findById(memberId)
+                .orElseThrow(() -> new NotExistedValueException("존재하지 않는 멤버 입니다"));
```
find와 get에 대한 관습에 대해서 학습해보세요.

https://tuhrig.de/find-vs-get/

RaZel713 replied:
학습 후 해당 메서드들의 네이밍을 확인하여 변경하였습니다!

## find란?
말 그대로 무언가를 찾아보는 동작을 의미합니다. 
이때 원하는 결과가 없어도 그것이 정상적인 상황일 수 있습니다. 

**Example:** 
사용자가 책 제목으로 검색했는데 해당 책이 존재하지 않는다면, 이는 오타일 수도 있고 단순히 그 책이 없는 것일 수도 있습니다. 
이 경우, Optional.empty()를 반환하거나 빈 리스트를 반환하며, 호출자가 결과 유무를 직접 판단하고 처리해야 합니다.

## get 이란?
특정 엔티티가 반드시 존재해야 한다는 전제 하에 사용합니다. 
존재하지 않는다면 이는 시스템 내부적인 문제로 간주되며, JPA는 EntityNotFoundException과 같은 예외를 발생시킵니다. 

**Example:** 
사용자가 어떤 책을 주문했고, 이를 기반으로 송장을 발행하려 할 때 
해당 책이 존재하지 않는다면 이는 데이터의 무결성이 깨진 심각한 상황이기 때문에 예외를 통해 바로 알리는 것이 좋습니다.

## 정리 

find는 **없을 수도 있다는 가능성을 인정하고 사용하는 방식**이며, 
get은 **무조건 있어야 한다는 가정 하에 사용하는 방식**입니다. 

따라서 메서드 반환 타입도 
find는 Optional<T> 또는 List<T> 형태로 처리하고, 
get은 T 타입을 직접 반환하면서 존재하지 않으면 예외를 던지는 방식으로 구현합니다.

## 📌 주의할 점
어떤 경우에도 null을 반환하는 방식은 피하고, 반환 값이 없을 수 있는 상황에서는 Optional.empty()나 빈 리스트를 사용해 명시적으로 의도를 드러내는 것이 더 안전하고 명확합니다.

jinyoungchoi95 commented on src/main/java/roomescape/config/WebMvcConfiguration.java:
```diff
@@ -0,0 +1,34 @@
+import roomescape.auth.interceptor.CheckLoginInterceptor;
+import roomescape.auth.resolver.LoginMemberArgumentResolver;
+import roomescape.auth.service.AuthService;
+
+
```
```suggestion

```

jinyoungchoi95 commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -0,0 +1,114 @@
+        this.password = password;
+    }
+
+    public Member(final String name, final String email, final String password) {
+        this(NON_SAVED_STATUS, Role.USER, name, email, password);
+    }
```
생성자에는 이름이 없으므로 어떤 값을 생성할 수 있을 지 모릅니다. `Role.USER`인 것을 임의로 생성자에 넣어버리면 의미 전달이 명확한 것 맞을까요? 정적 팩토리 메서드로 어떤 role의 member를 생성하는지 이름을 명확하게 전달하여야하지 않을까요?

RaZel713 replied:
``` java
 public static Member registerUser(final String name, final String email, final String password) {
        return new Member(NON_SAVED_STATUS, Role.USER, name, email, password);
}
```
사용자 회원 가입시 Member 생성에 있어, 
사용자 권한을 지닌 Member를 생성한다는 의미를 포함 시키기 위하여 
`registerUser`라는 정적 팩토리 메서드를 활용하도록 수정하였습니다!

jinyoungchoi95 commented on src/main/resources/application.properties:
```diff
@@ -3,3 +3,6 @@ spring.h2.console.path=/h2-console
 spring.datasource.username=test
 spring.datasource.password=
+security.jwt.token.secret-key=Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=
+# 1?? (1000 * 60 * 60 ???)
```
> 1?? (1000 * 60 * 60 ???)

utf-8 인코딩에 대해서 학습해보세요.

RaZel713 replied:
- **Unicode**: 전 세계의 모든 문자를 하나의 체계로 표현할 수 있도록 한 표준입니다. 하나의 문자에 고유 번호(코드 포인트)를 부여한다. 현대 문자를 표현할 때 가장 많이 사용되는 표준 문자 집합입니다.
- **UTF-8**: 유니코드를 구현한 인코딩 방식 중 하나입니다. 가변 길이 인코딩 방식이며, ASCII와 호환되고, 영어는 1바이트, 한글은 3바이트로 표현됩니다.
 
UTF-8 인코딩은 한글 한 글자를 2~3바이트로 표현합니다.
반면에 ISO-8859-1은 1바이트 문자셋으로, 한글을 표현할 수 없습니다.
한글이 포함된 UTF-8 파일을 ISO-8859-1로 읽으면, 한글의 바이트 조합을 각각 다른 문자로 잘못 해석합니다.
그래서 실제 한글이 물음표(?), �, 혹은 기타 깨진 글자(garbage characters) 로 보이게 됩니다.

➡️ Default encoding for properties files 설정을 제대로 확인하지 못했었습니다... 초기에 꼭 설정 확인하도록 하겠습니다..🥲

jinyoungchoi95 commented on src/main/resources/schema.sql:
```diff
@@ -14,14 +14,25 @@ CREATE TABLE theme
+CREATE TABLE member
+(
+    id       BIGINT                 NOT NULL AUTO_INCREMENT,
+    role     ENUM ('ADMIN', 'USER') NOT NULL,
```
domain코드의 enum과 다르게 db의 enum은 어떤 문제가 발생할까요?

RaZel713 replied:
### 📌 db enum 사용의 문제점

- db enum은 고정되어 있기 때문에 새로운 값을 추가하거나 변경하려면  sql문을 수정해 주어야 합니다.
- 도메인 코드에 새로운 role을 추가했지만, db에 추가한다면 런타임 에러가 발생합니다.

위와 같은 문제 발생으로 확장성이 저하되고 유지 보수가 어려워 집니다. 
db에서 enum 대신 varchar를 사용하도록 수정하였습니다!

jinyoungchoi95 commented on src/test/java/roomescape/auth/jwt/JwtTokenExtractorTest.java:
```diff
@@ -0,0 +1,84 @@
+
+class JwtTokenExtractorTest {
+
+    private JwtTokenExtractor jwtTokenExtractor;
+    private Key key;
+
+    @BeforeEach
+    void setUp() {
+        key = Keys.secretKeyFor(SignatureAlgorithm.HS256);
+        String base64SecretKey = Base64.getEncoder().encodeToString(key.getEncoded());
+        jwtTokenExtractor = new JwtTokenExtractor(base64SecretKey);
+    }
```
```suggestion
    private final Key key = Keys.secretKeyFor(SignatureAlgorithm.HS256);
    private final JwtTokenExtractor jwtTokenExtractor = new JwtTokenExtractor(
            Base64.getEncoder().encodeToString(key.getEncoded())
    );
```

junit5부터는 테스트의 라이프사이클이 메서드당이기 때문에 setUp을 할 필요가 없습니다.

https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-instance-lifecycle

RaZel713 replied:
setUp을 제거하였습니다!

알려주셔서 감사합니다 😊

jinyoungchoi95 commented on src/test/java/roomescape/auth/service/AuthServiceTest.java:
```diff
@@ -0,0 +1,92 @@
+        Member result = authService.findMemberByToken(token);
+
+        // then
+        assertThat(result.getId()).isEqualTo(1L);
+        assertThat(result.getEmail()).isEqualTo("admin@email.com");
```
`assertAll`이어야겠네요.

jinyoungchoi95 commented on src/test/java/roomescape/auth/service/AuthServiceTest.java:
```diff
@@ -0,0 +1,92 @@
+                .hasMessage("로그인 토큰이 존재하지 않습니다");
+    }
+
+    @DisplayName("토큰이 ADMIN 역할을 가진 경우 true를 반환한다")
```
admin이 아닐때의 테스트도 필요하지 않을까요?

RaZel713 replied:
admin이 아닐 때의 테스트를 추가하였습니다.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 라젤 😄

몇가지 코멘트 남겨두었으니 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

RaZel713 committed:
refactor: 정적 팩토리 메서드를 활용

RaZel713 committed:
test: USER 토큰으로 접속 시, 권한 판별 테스트 추가

RaZel713 committed:
test: 불필요한 setUp 메서드 제거

RaZel713 committed:
feat: 인증 예외를 추가해 상황에 맞게 인증, 인가 예외를 분리

RaZel713 committed:
refactor: schema의 member 테이블 생성시, role을 enum이 아닌, varchar로 생성

RaZel713 committed:
refactor: 불필요한 메서드 래핑 제거

RaZel713 committed:
refactor: 어드민 페이지 유저 계정 접속 테스트 추가

RaZel713 committed:
refactor: 주석 수정

RaZel713 committed:
refactor: 'jjwt' 대신 'auth0/java-jwt' 라이브러리를 사용

RaZel713 committed:
refactor: find와 get의 관습에 따른 메서드 네이밍 변경

RaZel713 reviewed (commented):
안녕하세요 ~ 오리!

리뷰 주신 부분을 학습하고 반영해보았습니다!
좀 더 빠르게 반영해야 했는데, 생각보다 정리하는데 시간이 걸렸습니다 🙏
페어 프로그래밍 기간 때문에 시간을 많이 쓸 수 없어 금요일에서야 리뷰 요청 드립니다. 

부족한 부분이 있거나 추가 학습하면 좋을 부분이 있으면 알려주시면 좋을 것 같습니다!👍

jinyoungchoi95 reviewed (approved):
안녕하세요 라젤 😄

2단계 리뷰요청 머지시간이 지나 머지하겠습니다.
방탈출 사용자 예약 미션기간동안 주고받았던 피드백이 앞으로도 많은 도움이 되길 바랍니다 😄

감사합니다 🙇‍♂️
